libs/sdk-utils/core/src/lib/utils.ts

```diff
@@ -0,0 +1,23 @@
+import { Provider } from '@ethersproject/abstract-provider';
+import { Contract, ContractInterface, ethers, Signer } from 'ethers';
+
+/**
+ * Casting a contract inline can lose type signature information.
+ * This helper adds a generic to properly ensure type information is not lost at compile time.
+ * @returns an ethers JS contract with full type inference
+ */
+export const typesafeContract = <T extends Contract>(
```

Comment:
How do you make sure that the `Contract` that's returned will indeed correspond to the type `T`? It is my understanding that typechain takes care of this, so that we don't need to deal with `abi` files. Did something change?

Reply:
Good question, got me thinking...

...I dug into the typechain examples for ethers and it seems you can directly create a contract using the typechain factories like so:

```ts
  const dai = Dai__factory.connect(DAI_ADDRESS, provider)
 ```

The advantage here is, as you said, there is no risk that the passed ABI is incorrect.

The disadvantage for the user is that the factory class must be available at runtime. 

The typesafe contract function assumes that the generic corresponds to the ABI. The ABI itself has no checking, fundamentally, we need a compilation step before we can interpret the ABI. 

I think a nice improvement here would be to allow the passing of the factory if it exists, or the ABI if not.

Reply:
> allow the passing of the factory if it exists, or the ABI if not.

This is a great idea!